### PR TITLE
Added edge case behavior testing for math bruteforce

### PR DESCRIPTION
**Pull request by shajder**
Fixes #2641

### CODE REVIEW

File: test_conformance/math_brute_force/edge_cases.cpp [hunk 1]
```diff
@@ -0,0 +1,708 @@
...
+            }
+        }
+    }
+
+    return err == -1 ? TEST_FAIL : TEST_PASS;
+}
+
+template <typename T> AnyValue abstract_to_anyvalue(const AbstractValue &av)
+{
+    // Int jest niezależny od precyzji — zawsze cl_int
```
Comment (svenvh) [inline test_conformance/math_brute_force/edge_cases.cpp]: Suggesting to translate this comment to English. :)
(and avoid the non-ASCII em dash)

File: test_conformance/math_brute_force/edge_cases.cpp [hunk 2]
```diff
@@ -0,0 +1,708 @@
...
+    std::size_t byte_size = 0;
+    std::array<uint8_t, CL_VALUE_MAX_BYTES> data{};
+
+    template <typename T> static AnyValue make(const T &val)
+    {
+        static_assert(sizeof(T) <= CL_VALUE_MAX_BYTES,
+                      "CLValue: type exceeds CL_VALUE_MAX_BYTES");
+        AnyValue v;
+        v.cl_type = "half";
+        if constexpr (std::is_same_v<T, cl_float>)
```
Comment (svenvh) [inline test_conformance/math_brute_force/edge_cases.cpp:50]: Needs a `#include <type_traits>`.

File: test_conformance/math_brute_force/edge_cases.cpp [hunk 3]
```diff
@@ -0,0 +1,708 @@
...
+const AbstractValue TWO = AV_F(2.0);
+const AbstractValue NEG_TWO = AV_F(-2.0);
+
+struct AbstractEdgeCase
+{
+    const char *func_name;
+    std::vector<AbstractValue> inputs;
+    AbstractValue expected;
+
+    bool expect_nan = false; // check isnan() instead of memcmp
```
Comment (svenvh) [inline test_conformance/math_brute_force/edge_cases.cpp]: where is `isnan` being called?

File: test_conformance/math_brute_force/edge_cases.cpp [hunk 4]
```diff
@@ -0,0 +1,708 @@
...
+        ec.inputs.push_back(abstract_to_anyvalue<T>(av));
+    ec.func_name = aec.func_name;
+    return ec;
+}
+
+inline cl_int run_edge_cases(const AbstractEdgeCase *cases, std::size_t count,
+                             cl_context context, cl_command_queue queue)
+{
+    cl_int overall = CL_SUCCESS;
+    log_info("float\n");
```
Comment (svenvh) [inline test_conformance/math_brute_force/edge_cases.cpp]: We should probably gate this on `gTestFloat` so that it can be toggled with `-f` like all other tests.

File: test_conformance/math_brute_force/edge_cases.cpp [hunk 5]
```diff
@@ -0,0 +1,708 @@
...
+    AbstractValue expected;
+
+    bool expect_nan = false; // check isnan() instead of memcmp
+
+    bool requires_inf_nan = false; // CL_FP_INF_NAN
+    bool requires_denorm = false; // CL_FP_DENORM
+    bool requires_rte = false; // CL_FP_ROUND_TO_NEAREST
+};
+
+static const AbstractEdgeCase edge_case_table[] = {
```
Comment (svenvh) [inline test_conformance/math_brute_force/edge_cases.cpp:174]: Maybe worth adding something like:

```suggestion
// Taken from OpenCL C Section 7.5.1. Additional Requirements Beyond ...
static const AbstractEdgeCase edge_case_table[] = {
```

File: test_conformance/math_brute_force/edge_cases.cpp [hunk 6]
```diff
@@ -0,0 +1,708 @@
...
+            AnyValue got;
+            got.byte_size = ec.expected.byte_size;
+            std::memcpy(got.data.data(), result.data(), got.byte_size);
+            if (!got.all_elements_nan<T>())
+            {
+                log_error("FAIL: %s(", ec.func_name);
+                for (std::size_t i = 0; i < ec.inputs.size(); ++i)
+                {
+                    if (i) log_error(", ");
+                    log_error("0x");
```
Comment (svenvh) [inline test_conformance/math_brute_force/edge_cases.cpp]: Doesn't `log_anyvalue` already print the `0x`?
Please check other `log_anyvalue` calls too.

File: test_conformance/math_brute_force/edge_cases.cpp [hunk 7]
```diff
@@ -0,0 +1,708 @@
...
+                    log_anyvalue(ec.inputs[i]);
+                }
+                log_error(") — expected NaN, got 0x");
```
Comment (svenvh) [inline test_conformance/math_brute_force/edge_cases.cpp]:
```suggestion
                log_error(") - expected NaN, got 0x");
```

File: test_conformance/math_brute_force/edge_cases.cpp [hunk 8]
```diff
@@ -0,0 +1,708 @@
...
+}
+
+inline cl_int run_edge_cases(const AbstractEdgeCase *cases, std::size_t count,
+                             cl_context context, cl_command_queue queue)
+{
+    cl_int overall = CL_SUCCESS;
+    log_info("float\n");
+    for (std::size_t i = 0; i < count; ++i)
+    {
+        auto aec = cases[i];
```
Comment (svenvh) [inline test_conformance/math_brute_force/edge_cases.cpp]: Does this really need to be a copy?
Same for the half and double cases further down.

File: test_conformance/math_brute_force/edge_cases.cpp [hunk 9]
```diff
@@ -0,0 +1,708 @@
...
+    bool expect_nan = false; // check isnan() instead of memcmp
+
+    bool requires_inf_nan = false; // CL_FP_INF_NAN
+    bool requires_denorm = false; // CL_FP_DENORM
+    bool requires_rte = false; // CL_FP_ROUND_TO_NEAREST
+};
+
+static const AbstractEdgeCase edge_case_table[] = {
+
+    { "acospi", { ONE }, POS_ZERO },
```
Comment (svenvh) [inline test_conformance/math_brute_force/edge_cases.cpp:176]: This table only covers the cases listed in the OpenCL C specification as fas as I can tell?  Should we also check for the cases from Section F.9 of the C99 Specification?  I'd suggest keeping the OpenCL and C99 lists somehow distinct (i.e., not interleaved).

File: test_conformance/math_brute_force/edge_cases.cpp [hunk 10]
```diff
@@ -0,0 +1,708 @@
...
+    }
+
+    return overall;
+}
+
+} // anonymous namespace
+
+REGISTER_TEST(math_edge_cases)
+{
+    return run_edge_cases(
```
Comment (svenvh) [inline test_conformance/math_brute_force/edge_cases.cpp:721]: The `-l` flag has no effect for this test currently.

Maybe we should simply skip the edge case tests entirely if `gSkipCorrectnessTesting` is set.  Any sort of "link check" is already covered by the other individual tests.

File: test_conformance/math_brute_force/edge_cases.cpp [hunk 11]
```diff
@@ -0,0 +1,708 @@
...
+                std::memcpy(&bits, v.data.data() + off, 2);
+                log_error("0x%04" PRIx16, bits);
+                break;
+            }
+        }
+    }
+}
+
+template <typename T>
+inline cl_int run_edge_case(const EdgeCaseSpec &ec, cl_context context,
```
Comment (svenvh) [inline test_conformance/math_brute_force/edge_cases.cpp:386]: This function is pretty inefficient because it builds + enqueues a new kernel for every edge case.  I understand that we want to keep the test code simple, so I'm not asking for a change (yet), but would like to hear some other opinions.

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 1 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp]: Done

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 2 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp:50]: I wonder, shouldn't this be detected with CI/CD ? Line added

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 3 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp]: Removed in the development process when I added support for half and double, comment removed.

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 4 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp]: Done

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 5 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp:174]: Done

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 6 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp]: Done

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 7 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp]: Done

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 8 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp]: Done

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 9 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp:176]: Certainly we should support `half_` versions. As for the rest I would appreciate WG expertise.

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 10 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp:721]: Done

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 11 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp:386]: As mentioned in the issue, we can batch edge-case kernels into groups. This will make the code slightly less readable, so I’d appreciate your input on whether it’s the acceptable tradeoff. Thanks for your review @svenvh

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 8 above)
Comment (svenvh) [reply test_conformance/math_brute_force/edge_cases.cpp]: Can they be `const` references?

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 11 above)
Comment (svenvh) [reply test_conformance/math_brute_force/edge_cases.cpp]: I'd expect that per-builtin-function batching shouldn't make the test code too complicated, and it would give a good improvement (e.g. the 22 `atan2pi` tests will require just 1 kernel execution).  If we'll be adding more edge cases (as per my other comment about the C99 edge cases), I'd say it's worth considering.

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 8 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp]: Corrected

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 11 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp]: > 22 `atan2pi` tests will require just 1 kernel execution

I just pushed a correction, the least intrusive solution is to group kernels based on function name (equal number of kernel arguments). On my machine ~120ms decreased to ~20ms.

File: test_conformance/math_brute_force/edge_cases.cpp [hunk 12]
```diff
@@ -382,12 +358,52 @@ void log_anyvalue(const AnyValue &v)
...
+        src += "[";
+        src += std::to_string(ind);
+        src += "]";
+    }
+    src += ");";
+
+    batch_cases.push_back(ec);
+}
+
+template <typename T>
```
Comment (svenvh) [inline test_conformance/math_brute_force/edge_cases.cpp]: It would be nice to have a comment about the high-level strategy, now that we're batching executions together.

File: test_conformance/math_brute_force/edge_cases.cpp [hunk 13]
```diff
@@ -402,44 +418,59 @@ inline cl_int run_edge_case(const EdgeCaseSpec &ec, cl_context context,
...
 
-        err =
-            clEnqueueWriteBuffer(queue, buf, CL_TRUE, 0, ec.inputs[i].byte_size,
-                                 ec.inputs[i].data.data(), 0, nullptr, nullptr);
+        static std::vector<uint8_t> inData;
+        inData.resize(batch_cases.front().inputs[i].byte_size
+                      * batch_cases.size());
+
+        size_t byte_offset = 0;
+        for (auto &elem : batch_cases)
```
Comment (svenvh) [inline test_conformance/math_brute_force/edge_cases.cpp]: `const`?

File: test_conformance/math_brute_force/edge_cases.cpp [hunk 14]
```diff
@@ -591,43 +634,66 @@ template <typename T> EdgeCaseSpec make_edge_case(const AbstractEdgeCase &aec)
...
+    return ret;
+}
+
 inline cl_int run_edge_cases(const AbstractEdgeCase *cases, std::size_t count,
                              cl_context context, cl_command_queue queue)
 {
     cl_int overall = CL_SUCCESS;
     if (gTestFloat)
     {
         log_info("float test\n");
```
Comment (svenvh) [inline test_conformance/math_brute_force/edge_cases.cpp]: Remove ` test` to be consistent with the `double` and `half` cases.

File: test_conformance/math_brute_force/edge_cases.cpp [hunk 15]
```diff
@@ -461,64 +492,76 @@ inline cl_int run_edge_case(const EdgeCaseSpec &ec, cl_context context,
...
             log_error("ERROR: clEnqueueNDRangeKernel failed for '%s': %d\n",
-                      ec.func_name, err);
+                      batch_cases.front().func_name, err);
             return TEST_FAIL;
         }
     }
 
     {
-        std::array<uint8_t, CL_VALUE_MAX_BYTES> result{};
+        static std::vector<uint8_t> result;
```
Comment (svenvh) [inline test_conformance/math_brute_force/edge_cases.cpp]: Why static?

File: test_conformance/math_brute_force/edge_cases.cpp [hunk 16]
```diff
@@ -382,12 +358,52 @@ void log_anyvalue(const AnyValue &v)
...
+        src += " *out";
+        for (std::size_t i = 0; i < ec.inputs.size(); ++i)
+        {
+            src += ",\n    __global const ";
+            src += ec.inputs[i].cl_type;
+            src += " *in";
+            src += std::to_string(i);
+        }
+        src += ")\n{";
+    }
```
Comment (svenvh) [inline test_conformance/math_brute_force/edge_cases.cpp]: Perhaps worth having an else with `assert(batch_cases.front().inputs.size() == ec.inputs.size());` here?

File: test_conformance/math_brute_force/edge_cases.cpp [hunk 17]
```diff
@@ -402,44 +418,59 @@ inline cl_int run_edge_case(const EdgeCaseSpec &ec, cl_context context,
...
-                      ec.func_name, err);
+                      batch_cases.front().func_name, err);
             return TEST_FAIL;
         }
         in_bufs.push_back(buf);
 
-        err =
-            clEnqueueWriteBuffer(queue, buf, CL_TRUE, 0, ec.inputs[i].byte_size,
-                                 ec.inputs[i].data.data(), 0, nullptr, nullptr);
+        static std::vector<uint8_t> inData;
```
Comment (svenvh) [inline test_conformance/math_brute_force/edge_cases.cpp]: Why static?

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 11 above)
Comment (svenvh) [reply test_conformance/math_brute_force/edge_cases.cpp]: Thanks, that brings it down from 20 to 6 seconds in my (emulated) environment, so that's a good enough improvement for me.

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 15 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp]: To avoid continuous allocations/deallocations, the vector retains its capacity between calls so resize() does not reallocate as long as the size stays within the previously allocated capacity.

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 17 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp]: as above

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 12 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp]: Done

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 13 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp]: Done

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 14 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp]: Done

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 16 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp]: Done

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 9 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp:176]: moreover, [this](https://github.com/KhronosGroup/OpenCL-Docs/issues/1538) discussion may be relevant

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 15 above)
Comment (svenvh) [reply test_conformance/math_brute_force/edge_cases.cpp]: The potential benefit here looks marginal in the bigger scheme of things, and I don’t think it outweighs the readability/maintainability cost of introducing hidden persistent state.  So I would encourage dropping `static`.

Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp]: Done, working buffers turned into class members.

File: test_conformance/math_brute_force/edge_cases.cpp (hunk 17 above)
Comment (shajder) [reply test_conformance/math_brute_force/edge_cases.cpp]: Done

**Review by svenvh — APPROVED** [review]
LGTM, thanks!

There's still the open question about the C99 edge cases, but I would suggest merging the current state and adding more edge cases in a followup PR (potentially after ironing out the specification wrt listing edge cases).

**Comment by neiltrevett** [comment]
Merged as discussed in call on 5May26
